Gateways, allow for configuring a gatewaygroup as the default gateway.

PiBa-NL opened the pull request:
Gateways, allow for configuring a gatewaygroup as the default gateway.
-Avoid changing routes by just visiting a webgui page.
-Avoid change some unneeded events when nothing changed.
  
- [x] Redmine Issue: https://redmine.pfsense.org/issues/8187
- [x] Ready for review

jim-p commented:
I spoke with @PiBa-NL on IRC about this, there are a couple more tweaks he's going to make before we try testing it out.

rbgarga commented:
Good to know @jim-p. So I'll wait for it too. BTW, it's too big for 2.4 so I marked to be merged post 2.4.0 release

luckman212 commented:
@PiBa-NL I haven't looked at your code yet, but does this patch somehow fix https://redmine.pfsense.org/issues/5476 ? Enable traffic from the firewall itself to route using PBR?

PiBa-NL commented:
@luckman212 I don't touch the pf and its pbr functionality, though my patch is aimed at addressing the issues of pfSense not having dns and other functionality when its 'default' gateway is not available. It will pick a new default from the chosen gatewaygroup that is still 'up'. That doesnt make pbr work, as that would be configureable to behave different for different firewallrules matching traffic.. My patch only juggles with the default route..

jim-p commented:
@PiBa-NL Is this complete now, or is there still something else you planned to add/change?

PiBa-NL commented:
not complete yet.. ive got several more changes to add.. but need to clean them up a bit.. ill try and do that in next few days.

jim-p commented:
In an effort to improve the tracking of changes and bug/feature requests, we have decided to require an entry on the [pfSense Redmine issue tracker](https://redmine.pfsense.org) associated with every pull request, and likewise the Redmine entry should also have a link back to the pull request.

If you could, please associate this PR to a Redmine issue either by locating an existing issue at https://redmine.pfsense.org or by creating a new issue. Add a link to the Redmine issue that points to this PR, and also add a link on the PR to the Redmine issue.

For more information, see https://doc.pfsense.org/index.php/Submitting_a_Pull_Request_via_Github

Thanks!

jim-p commented:
Redmine issue: https://redmine.pfsense.org/issues/8187

jim-p commented on src/etc/inc/upgrade_config.inc:
```diff
@@ -5493,6 +5493,90 @@ function upgrade_172_to_173() {
 	}
 }
 
+function upgrade_173_to_174() {
```
This will need to change, it's already at 17.4 now, so this would be `upgrade_174_to_175()`.

jim-p commented on src/etc/inc/globals.inc:
```diff
@@ -74,7 +74,7 @@ $g = array(
 	"crashreporterurl" => "https://crashreporter.pfsense.org/crash_reporter.php",
 	"debug" => false,
-	"latest_config" => "17.3",
+	"latest_config" => "17.4",
```
This will need to be 17.5

PiBa-NL commented:
Okay bumped to 17.6 .

sbeaver-netgate commented:
I will do what I can to test and merge this week.

PiBa-NL commented:
Okay thanks, rebased it.

nqb commented:
Hello @sbeaver-netgate,

Have you test PR ? Do you think you can merge it for 2.4.4 ?
